client/virtual_channel_settlement.go

```diff
@@ -0,0 +1,188 @@
+// Copyright 2021 - See NOTICE file for copyright holders.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package client
+
+import (
+	"context"
+
+	"github.com/pkg/errors"
+	"perun.network/go-perun/channel"
+	"perun.network/go-perun/wire"
+)
+
+// withdrawVirtualChannel proposes to release the funds allocated to the
+// specified virtual channel.
+func (c *Channel) withdrawVirtualChannel(ctx context.Context, virtual *Channel) (err error) {
+	// Lock machine while update is in progress.
+	if !c.machMtx.TryLockCtx(ctx) {
+		return errors.Errorf("locking machine mutex in time: %v", ctx.Err())
+	}
+	defer c.machMtx.Unlock()
+
+	state := c.state().Clone()
+	state.Version++
+
+	virtualAlloc, ok := state.SubAlloc(virtual.ID())
+	if !ok {
+		c.Log().Panicf("sub-allocation %x not found", virtualAlloc.ID)
+	}
+
+	if !virtualAlloc.BalancesEqual(virtual.state().Allocation.Sum()) {
+		c.Log().Panic("sub-allocation does not equal accumulated sub-channel outcome")
+	}
+
+	virtualBalsRemapped := virtual.translateBalances(virtualAlloc.IndexMap)
+
+	// We assume that the asset types of parent channel and virtual channel are the same.
+	state.Allocation.Balances = state.Allocation.Balances.Add(virtualBalsRemapped)
+
+	if err := state.Allocation.RemoveSubAlloc(virtualAlloc); err != nil {
+		c.Log().WithError(err).Panicf("removing sub-allocation with id %x", virtualAlloc.ID)
+	}
+
+	err = c.updateGeneric(ctx, state, func(mcu *msgChannelUpdate) wire.Msg {
+		return &virtualChannelSettlementProposal{
+			msgChannelUpdate:     *mcu,
+			ChannelParams:        *virtual.Params(),
+			FinalState:           *virtual.state(),
+			FinalStateSignatures: virtual.machine.CurrentTX().Sigs,
+		}
+	})
+
+	return errors.WithMessage(err, "update parent channel")
+}
+
+func (c *Client) handleVirtualChannelSettlementProposal(
+	parent *Channel,
+	prop *virtualChannelSettlementProposal,
+	responder *UpdateResponder,
+) {
+	err := c.validateVirtualChannelSettlementProposal(parent, prop)
+	if err != nil {
+		c.rejectProposal(responder, err.Error())
+	}
+
+	ctx, cancel := context.WithTimeout(c.Ctx(), virtualChannelFundingTimeout)
```

Comment:
maybe rename `virtualChannelFundingTimeout` to something like `virtualChannelUpdateTimeout` because it is not only used for funding now.

Reply:
Couldn't find a good name for both cases, added a separate variable for settling.